avm/src/vm/memory.rs

```diff
@@ -136,67 +149,132 @@ impl HandlerMemory {
+  /// all the forked children drop their reference to the parent, we take
+  /// the parent out of the Arc after parallel work is done and then join on its children.
+  pub fn drop_parent(self: &mut HandlerMemory) {
+    if self.parent.is_some() {
+      let arc = self.parent.take();
+      drop(arc);
+    }
+  }
+
+  /// Returns true if the idxs are valid in self
+  fn is_idx_defined(self: &HandlerMemory, a: usize, b: usize) -> bool {
+    let is_raw = a == std::usize::MAX;
+    let safe_mem_space = self.mem_addr == 1 || a >= self.mem_addr;
+    let is_fixed = safe_mem_space && self.mems.len() > a && self.mems[a].len() > b;
+    let is_fractal = safe_mem_space && self.mems.len() > a && b == std::usize::MAX;
+    return is_raw || is_fixed || is_fractal;
+  }
+
+  /// Recursively finds the parent and returns None if the idxs
+  /// belong to self or otherwise a reference to the parent
+  fn hm_for_idxs(self: &HandlerMemory, a: usize, b: usize) -> Option<&Arc<HandlerMemory>> {
+    if self.is_idx_defined(a,b) {
+      return None;
+    }
+    let res = self.parent.as_ref().unwrap().hm_for_idxs(a, b);
+    return if res.is_none() { self.parent.as_ref() } else { res };
+  }
+
+  /// Takes a given address and looks up the fractal location and
+  /// `mems` indexes relevant to it if available
+  fn addr_to_idxs_opt(self: &HandlerMemory, addr: i64) -> Option<(usize, usize)> {
     return if addr >= 0 {
-     self.addr.0[addr as usize]
+      *self.addr.0.get(addr as usize).unwrap_or(&None)
     } else if addr <= CLOSURE_ARG_MEM_END {
-      self.addr.1[(addr - CLOSURE_ARG_MEM_START) as usize]
+      *self.addr.1.get((addr - CLOSURE_ARG_MEM_START) as usize).unwrap_or(&None)
     } else {
-      (0, ((-1 * addr - 1) / 8) as usize)
-    }
+      Some((0, ((-1 * addr - 1) / 8) as usize))
+    };
+  }
+
+  /// Recursively taks a given address and looks up the fractal location and
+  /// `mems` indexes relevant to it. It also returns an Option that is
+  /// None if the address is in self or a ptr to the ancestor if `self` is forked
+  fn addr_to_idxs(self: &HandlerMemory, addr: i64, ) -> ((usize, usize), Option<&Arc<HandlerMemory>>) {
+    let idxs = self.addr_to_idxs_opt(addr);
+    return if idxs.is_none() {
+      // fail if no parent
+      if self.parent.is_none() {
+        panic!("Memory address referenced in parent, but no parent pointer defined");
+      }
+      let res = self.parent.as_ref().unwrap().addr_to_idxs(addr);
+      let hm = if res.1.is_none() { self.parent.as_ref() } else { res.1 };
+      (res.0, hm)
+    } else {
+      let res = idxs.unwrap();
+      (res, self.hm_for_idxs(res.0, res.1))
+    };
   }
 
   /// Reads fixed data from a given address.
   pub fn read_fixed(self: &HandlerMemory, addr: i64) -> i64 {
-    let (a, b) = self.addr_to_idxs(addr);
+    let ((a, b), hm_opt) = self.addr_to_idxs(addr);
+    let hm = if hm_opt.is_none() { self } else { hm_opt.unwrap() };
```

Comment:
This line looks like it could just be in the `else` branch below on line 220. No need to check or unwrap the HM if the value we got back is a fixed value.

Reply:
Yes, good point!